### DISCUSSION TITLE
Using the NodeJS version, built from source

### DISCUSSION THREAD

**Discussion by tprelog** [zwave-js/zwave-js-ui#571] [Q&A]
Hi

I have some questions about running the NodeJS version, built from source. But first, a little background about why I'm asking

I've installed zwavejs2mqtt in a jail on my TrueNAS CORE server - So far everything seems to working great. Since the OS is based on FreeBSD, I need to follow the instructions to install the [NodeJS or PKG version](https://zwave-js.github.io/zwavejs2mqtt/#/getting-started/quick-start?id=nodejs-or-pkg-version) - The package version is not complied for FreeBSD, so I need to build the project from source.

Rather than building the code downloaded using git, I have modified the package version command, to download the source code for the latest tagged release instead.

```bash
tarball_url="$(curl -s https://api.github.com/repos/zwave-js/zwavejs2mqtt/releases/latest | grep tarball_url | cut -d : -f 2,3 | tr -d \",)"
```

**First question**

When I compile the code from `source.tar.gz`, I see the following error - I assume this is because I did not `git clone` the repository. Is this going to cause an issue or can it be ignored?

```bash
> husky@4.3.8 install /tmp/build_src.jRE0FTWH/zwave-js-zwavejs2mqtt-301f4a0/node_modules/@zwave-js/server/node_modules/husky
> node husky install

husky > Setting up git hooks
git rev-parse command failed. Got fatal: not a git repository (or any of the parent directories): .git
husky > Failed to install
```

**Second question**

If the answer to the first question is `It's going to cause an issue`  --  When I build the project _in a directory I cloned, husky does not fail_ to install. Is there a magic git command, that will always clone the latest tagged release?

**Third question**

Finally, when a new version is released and it's time to update...  

A little more background - So to make life a little easier, I basically copied the [Dockerfile](https://github.com/zwave-js/zwavejs2mqtt/blob/master/docker/Dockerfile) into a bash script, that compiles the downloaded source code, builds the project in a temp directory, then copies it to an "installed" location.

I should also point out, I set `STORE_DIR` to location outside of the "installed" directory.

...now back to the question. When `version 1.1.1` was released, I took the following steps to upgrade.

1. download / compile latest tagged release (`1.1.1`) in a temp directory
2. deleted the existing version in the "installed" location
3.  copied latest release from temp directory to "installed" location
4. restart zwavejs2mqtt
5. profit!

Can you say if these steps seem appropriate, or can you suggest an alternative steps for upgrades?

Thanks

**Comment by robertsLando** [comment]
@tprelog Sorry but why don't simply use a docker file and do a `docker pull` instead of building your own?

**Reply by tprelog** [reply]
I am <s>trying to run</s> running Z-Wave JS to MQTT from my [TrueNAS CORE](https://www.truenas.com/truenas-core/) server, which is based on FreeBSD - Unfortunately there is not a working version of Docker available for FreeBSD at this time.

Building from source is the only way I can  keep this my NAS.

PS -- I made this a TrueNAS CORE Community Plugin. I hope you don't mind - I'll have it removed if you do

![image](https://user-images.githubusercontent.com/35702532/107952231-7c244b00-6f67-11eb-96b7-6150d7e818a9.png)

**Reply by robertsLando** [reply]
@tprelog No worries it's ok for me, thanks for that :)

**Reply by tprelog** [reply]
Update on... running an update

For the record, using the same update procedure described in the OP, I have upgrade from `1.1.1` to `1.2.1` - This time there were no errors - I'm not sure if it was intentional to fix the previously mentioned husky error, but it does seem to be resolved

**Reply by robertsLando** [reply]
@tprelog The problem was from zwavejs server, seems that for some reason that package was released with an npmshrinkwrap file, this caused that on every install we were adding ALL the dev packages (including husky) of that. Fixed on their side with this pr: https://github.com/zwave-js/zwave-js-server/pull/114

**Reply by tprelog** [reply]
Thanks explaining @robertsLando

I have another question, if I was smart this would have been my first question :roll_eyes:

Anyways - **What version of Node do you recommend using ?**

I have been using Node 15 - `v15.8.0` to be exact - for a few weeks without any noticeable issues - Do you think it's OK or would you recommend another version? I have these versions also available `node10` `node12` and `node14`

**Reply by robertsLando** [reply]
I always raccomend to use last LTS version, in this case node14

**Reply by tprelog** [reply]
Hi, I'm back again....

I was looking at `updateDevices` [from the Dockerfile.](https://github.com/zwave-js/zwavejs2mqtt/blob/master/docker/Dockerfile#L21)

Seems self explanatory so I added it to my TrueNAS Plugin's [install script](https://github.com/tprelog/iocage-zwavejs2mqtt/blob/master/overlay/root/bin/install_zj2m#L24) -- I figure this is for someone who needs to build with the latest zwavejs device config to either test changes, or if they just can not wait for the next zwave2mqtt release

I've already tried it, and _seems_ to work. **I wanted to ask if there is any danger in doing this?**


BTW, thanks for taking to time to respond and explain things. I don't mean to take your time away from officially supported installation methods for this - If I am being to needy, please let me know and I will stop :smiley:

**Reply by robertsLando** [reply]
> I've already tried it, and seems to work. I wanted to ask if there is any danger in doing this?

The only danger would be if there will be some kind of breaking change in future but I think it's something that's difficult to happen so I wouldn't woorry too much :)
